Internal Compiler Error: CUDA in Debug mode

Concretely, that's my cmake setup:
```bash
cmake -DAMReX_SPACEDIM=3 -DAMReX_GPU_BACKEND=CUDA -DAMReX_HDF5=YES -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/home/install_path/amrex_cuda_3D_debug -DBUILD_SHARED_LIBS=ON -DAMReX_PROBINIT=NO ..
```

The error appears at 8%
```
...
[  8%] Building CUDA object Src/CMakeFiles/amrex.dir/Base/AMReX_DistributionMapping.cpp.o
Segmentation fault (core dumped)
make[2]: *** [Src/CMakeFiles/amrex.dir/build.make:202: Src/CMakeFiles/amrex.dir/Base/AMReX_DistributionMapping.cpp.o] Error 139
make[1]: *** [CMakeFiles/Makefile2:232: Src/CMakeFiles/amrex.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```
I am using cuda 12.1

Thanks, can confirm as well now, still on GNU 11.3.0 + NVCC 12.1.105 (Ubuntu 22.04):
```console
$ cmake -S . -B build -DCMAKE_BUILD_TYPE=Debug -DAMReX_GPU_BACKEND=CUDA

$ cmake --build build --verbose
...
/usr/bin/cmake -S/home/axel/src/amrex -B/home/axel/src/amrex/build --check-build-system CMakeFiles/Makefile.cmake 0
/usr/bin/cmake -E cmake_progress_start /home/axel/src/amrex/build/CMakeFiles /home/axel/src/amrex/build//CMakeFiles/progress.marks
/usr/bin/gmake  -f CMakeFiles/Makefile2 all
gmake[1]: Entering directory '/home/axel/src/amrex/build'
/usr/bin/gmake  -f Src/CMakeFiles/amrex.dir/build.make Src/CMakeFiles/amrex.dir/depend
gmake[2]: Entering directory '/home/axel/src/amrex/build'
cd /home/axel/src/amrex/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /home/axel/src/amrex /home/axel/src/amrex/Src /home/axel/src/amrex/build /home/axel/src/amrex/build/Src /home/axel/src/amrex/build/Src/CMakeFiles/amrex.dir/DependInfo.cmake --color=
Dependencies file "Src/CMakeFiles/amrex.dir/Base/AMReX_DistributionMapping.cpp.o.d" is newer than depends file "/home/axel/src/amrex/build/Src/CMakeFiles/amrex.dir/compiler_depend.internal".
Consolidate compiler generated dependencies of target amrex
gmake[2]: Leaving directory '/home/axel/src/amrex/build'
/usr/bin/gmake  -f Src/CMakeFiles/amrex.dir/build.make Src/CMakeFiles/amrex.dir/build
gmake[2]: Entering directory '/home/axel/src/amrex/build'
[  1%] Building CUDA object Src/CMakeFiles/amrex.dir/Base/AMReX_DistributionMapping.cpp.o
cd /home/axel/src/amrex/build/Src && /usr/local/cuda/bin/nvcc -forward-unknown-to-host-compiler  -I/home/axel/src/amrex/Src/Base -I/home/axel/src/amrex/Src/Base/Parser -I/home/axel/src/amrex/Src/Boundary -I/home/axel/src/amrex/Src/AmrCore -I/home/axel/src/amrex/Src/Amr -I/home/axel/src/amrex/Src/LinearSolvers/MLMG -I/home/axel/src/amrex/Src/LinearSolvers/OpenBC -I/home/axel/src/amrex/build -isystem=/usr/lib/x86_64-linux-gnu/openmpi/include -isystem=/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi -isystem=/usr/local/cuda/include -g --generate-code=arch=compute_86,code=[compute_86,sm_86] --expt-relaxed-constexpr --expt-extended-lambda -Xcudafe --diag_suppress=esa_on_defaulted_function_ignored -maxrregcount=255 -Xcudafe --display_error_number --use_fast_math -G --source-in-ptx -Xcompiler -rdynamic "-Xcompiler=-O0 -ggdb -Wall -Wno-sign-compare -Wno-unused-but-set-variable -Werror=return-type" -MD -MT Src/CMakeFiles/amrex.dir/Base/AMReX_DistributionMapping.cpp.o -MF CMakeFiles/amrex.dir/Base/AMReX_DistributionMapping.cpp.o.d -x cu -dc /home/axel/src/amrex/Src/Base/AMReX_DistributionMapping.cpp -o CMakeFiles/amrex.dir/Base/AMReX_DistributionMapping.cpp.o
Segmentation fault (core dumped)
gmake[2]: *** [Src/CMakeFiles/amrex.dir/build.make:202: Src/CMakeFiles/amrex.dir/Base/AMReX_DistributionMapping.cpp.o] Error 139
gmake[2]: Leaving directory '/home/axel/src/amrex/build'
gmake[1]: *** [CMakeFiles/Makefile2:214: Src/CMakeFiles/amrex.dir/all] Error 2
gmake[1]: Leaving directory '/home/axel/src/amrex/build'
gmake: *** [Makefile:136: all] Error 2
```

We add `--source-in-ptx` manually in these files can can disable it:
- `Tools/CMake/AMReX_SetupCUDA.cmake`
- `Tools/CMake/AMReXParallelBackends.cmake`

We add the flag for the `AMReX_CUDA_SHOW_CODELINES` option being true. It is set in `Tools/CMake/AMReXCUDAOptions.cmake` and is `ON` by default.

--- a/Tools/CMake/AMReXCUDAOptions.cmake
+++ b/Tools/CMake/AMReXCUDAOptions.cmake
@@ -85,8 +85,16 @@ cuda_print_option(AMReX_CUDA_DEBUG)
 
 # both are performance-neutral debug symbols
 option(AMReX_CUDA_SHOW_LINENUMBERS "Generate line-number information (optimizations: on)" ON)
-option(AMReX_CUDA_SHOW_CODELINES "Generate source information in PTX (optimizations: on)" ON)
 cuda_print_option(AMReX_CUDA_SHOW_LINENUMBERS)
+
+# https://github.com/AMReX-Codes/amrex/issues/3215
+# Nvidia Bug ID: 4088095
+if (CMAKE_CUDA_COMPILER_VERSION VERSION_GREATER_EQUAL 12.1)
+    set(AMReX_CUDA_SHOW_CODELINES_DEFAULT OFF)
+else()
+    set(AMReX_CUDA_SHOW_CODELINES_DEFAULT ON)
+endif()
+option(AMReX_CUDA_SHOW_CODELINES "Generate source information in PTX (optimizations: on)" AMReX_CUDA_SHOW_CODELINES_DEFAULT)
 cuda_print_option(AMReX_CUDA_SHOW_CODELINES)
 
 option(AMReX_CUDA_BACKTRACE "Generate host function symbol names (better cuda-memcheck)" ${AMReX_CUDA_DEBUG})

